metals/src/main/scala/scala/meta/internal/metals/Refactorings.scala

```diff
@@ -0,0 +1,128 @@
+package scala.meta.internal.metals
+
+import scala.concurrent.{Future, ExecutionContext}
+import scala.meta._
+import scala.meta.pc.CancelToken
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags._
+import scala.meta.internal.semanticdb.{SymbolOccurrence, TextDocument}
+import org.eclipse.{lsp4j => l}
+
+trait Refactoring {
+  def contribute(
+      params: l.CodeActionParams,
+      trees: Trees,
+      buffers: Buffers,
+      semanticdbs: Semanticdbs,
+      symbolSearch: MetalsSymbolSearch,
+      token: CancelToken
+  )(implicit ec: ExecutionContext): Future[Seq[l.CodeAction]]
+}
+
+object Refactoring {
+
+  object UseNamedArguments extends Refactoring {
+
+    override def contribute(
+        params: l.CodeActionParams,
+        trees: Trees,
+        buffers: Buffers,
+        semanticdbs: Semanticdbs,
+        symbolSearch: MetalsSymbolSearch,
+        token: CancelToken
+    )(implicit ec: ExecutionContext): Future[Seq[l.CodeAction]] = {
+      scribe.debug("Running contribute for UseNamedArguments")
+
+      def findMethodApplyTreeUnderCursor(
+          root: Tree,
+          range: Position
+      ): Option[Term.Apply] =
+        root
+          .collect {
+            case t @ Term.Apply(_, _)
+                if t.pos.start <= range.start && t.pos.end >= range.end =>
+              t
+          }
+          .sortBy(_.pos.start)
+          .lastOption
+
+      def findSymbolTree(tree: Tree): Term.Name = tree match {
+        case x @ Term.Name(_) => x
+        case Term.Select(t, x) => x
+        case Term.Apply(x, _) => findSymbolTree(x)
+      }
+
+      def findSymbolOccurrence(
+          textDocument: TextDocument,
+          symbolTreePos: Position
+      ): Option[SymbolOccurrence] =
+        textDocument.occurrences.find(so =>
+          so.getRange.startLine == symbolTreePos.startLine &&
+            so.getRange.startCharacter == symbolTreePos.startColumn
+        )
+
+      def buildEdits(
+          methodApplyTree: Term.Apply,
+          paramNames: List[String]
+      ): List[l.TextEdit] = {
+        methodApplyTree.args.zip(paramNames).flatMap {
+          case (Term.Assign(_, _), _) =>
+            // already a named argument, no edit needed
+            None
+          case (term, paramName) =>
+            val position =
```

Comment:
The position here might need to be adjusted using EditDistance - you can get it from the `DefinitionProvider.positionOccurence`

Reply:
I don't think we need to do any adjustment here? Because we are getting the offsets from a tree given to us by `Trees`, which uses `Buffers` to take account of unsaved edits.

I'll try to write a test for files with unsaved edits to confirm this. That's the only test I haven't been able to write yet.

Reply:
Right! I think it should be fine then. This is being tested for example in https://github.com/scalameta/metals/blob/2475a952a4eaf7f3b5e41246e091c3757f244908/tests/unit/src/test/scala/tests/RenameLspSuite.scala#L604-L609
but this test doesn't need to be added inside of this PR I think.

Reply:
I've confirmed via manual testing in VSCode that the code action does the right thing when the file has unsaved edits. I tried to write a unit test for it, but after 2 hours of battling I only succeeded in writing a test that proves the opposite 😢 

I'll open an issue for writing the test.